state/protocol/util.go

```diff
@@ -69,3 +70,26 @@ func IsSporkRootSnapshot(snapshot Snapshot) (bool, error) {
 	}
 	return true, nil
 }
+
+// FindGuarantors decodes the signer indices from the guarantee, and finds the guarantor identifiers from protocol state
+func FindGuarantors(state State, guarantee *flow.CollectionGuarantee) ([]flow.Identifier, error) {
+	snapshot := state.AtBlockID(guarantee.ReferenceBlockID)
+	epochs := snapshot.Epochs()
+	epoch := epochs.Current()
+	cluster, err := epoch.ClusterByChainID(guarantee.ChainID)
+
+	if err != nil {
+		// protocol state must have validated the block that contains the guarantee, so the cluster
+		// must be found, otherwise, it's an internal error
+		return nil, fmt.Errorf(
+			"internal error retrieving collector clusters for guarantee (ReferenceBlockID: %v, ChainID: %v): %w",
+			guarantee.ReferenceBlockID, guarantee.ChainID, err)
+	}
+
+	guarantorIDs, err := signature.DecodeSignerIndicesToIdentifiers(cluster.Members().NodeIDs(), guarantee.SignerIndices)
+	if err != nil {
+		return nil, fmt.Errorf("could not decode guarantor indices: %w", err)
```

Comment:
This change caused a cycle reference, since we've indicated in comments that the type of the returned error . I think it's the caller's responsibility to how to distinguish the error and handle them properly.